src/collapse/common.rs

```diff
@@ -0,0 +1,780 @@
+use std::borrow::Cow;
+use std::io;
+use std::mem;
+use std::sync::Arc;
+
+use chashmap::CHashMap;
+use crossbeam::channel;
+use fnv::FnvHashMap;
+use lazy_static::lazy_static;
+
+const CAPACITY_HASHMAP: usize = 512;
+pub(crate) const CAPACITY_READER: usize = 128 * 1024;
+pub(crate) const DEFAULT_NSTACKS_PER_JOB: usize = 100;
+const NBYTES_PER_STACK_GUESS: usize = 1024;
+const RUST_HASH_LENGTH: usize = 17;
+
+lazy_static! {
+    #[doc(hidden)]
+    pub static ref DEFAULT_NTHREADS: usize = num_cpus::get();
+}
+
+/// Private trait for internal library authors.
+///
+/// If you implement this trait, your type will implement the public-facing
+/// `Collapse` trait as well. Implementing this trait gives you parallelism
+/// for free as long as you adhere to the requirements described in the
+/// comments below.
+pub trait CollapsePrivate: Clone + Send + Sized + Sized {
+    // *********************************************************** //
+    // ********************* REQUIRED METHODS ******************** //
+    // *********************************************************** //
+
+    /// Some formats, such as `dtrace`, contain a header or other non-stack
+    /// information at the beginning of their input files. If header information
+    /// is present, this method **must** consume it (i.e. advance the provided
+    /// reader past it).
+    ///
+    /// This method also provides an opportunity to do processing of actual
+    /// stack data on the main thread before worker threads are spun up. For
+    /// an example of why this might be necessary, see `perf`, which can require
+    /// reading the first stack in order to know how to process the rest.
+    ///
+    /// If the format you are working with does not contain header information
+    /// or does not need any special, up-front processing, just have this method
+    /// return `Ok(())` immediately.
+    fn pre_process<R>(&mut self, reader: &mut R, occurrences: &mut Occurrences) -> io::Result<()>
+    where
+        R: io::BufRead;
+
+    /// The primary method.
+    ///
+    /// It receives a reader whose header has already been consumed (see above), as
+    /// well as a mutable reference to an `Occurences` instance (just a hashamp that
+    /// works across multiple threads). Implementators should parse the stack data
+    /// contained in the reader and write output to the provided `Occurrences` map.
+    ///
+    /// So that this method may be called multiple times, all internal
+    /// state contained in `self` (e.g. any stack buffers or the like) **must** be
+    /// reset before this method returns. Also, this method may **not** use threads.
```

Comment:
This "reset" business still strikes me as weird. I _think_ all this is getting at is that it should treat the end of the input correctly? What state is there to reset beyond what you would normally reset when the input has ended (namely to end the current stack)?

Reply:
This is the property that must hold. How would you prefer to word it?
```rust
let mut collapser = <some collapser>;
let input1 = <some stuff>;
let input2 = input1.clone();
let mut occurrences1 = Occurrences::new(<some number>);
let mut occurrences2 = occurrences1.clone();
collapser.collapse_single_threaded(input1, &mut occurrences1)?;
collapser.collapse_single_threaded(input2, &mut occurrences2)?;
assert_eq!(occurrences1, occurrences2);
```

In quasi-English: same input + same fresh occurrences map -> must produce the same effects on occurrences.

Reply:
Well, for example, the current phrasing implies that you're not allowed to keep using the same caching structures for string interning across executions, even though I don't think there's a reason to require that. I think giving that exact code segment in an `ignore` code block would go a long way. Maybe also something like:

> This method may be called multiple times to process batches of incoming samples. Therefore, make sure that when end-of-file is reached, the collapser now considers itself back at the top-level context (e.g., not in the middle of a stack).

Reply:
I like that phrasing.

Reply:
Unfortunately, I don't think I can add the code sample because I can't access anything private in a doc test. I can't import `Occurrences` for example. And not being able to do that makes the compiler yell at you when running `cargo test`.